### PR TITLE
add configuration to build a dynamic library

### PR DESCRIPTION
**Pull request by folkertdev**

### CODE REVIEW

**Comment by folkertdev** [comment]
still needs some CI checks I think (for e.g. bzpipe working)

File: libbzip2-rs-sys-cdylib/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+readme = "README.md"
+license = "Zlib"
+repository = "https://github.com/trifectatechfoundation/libbzip2-rs"
+homepage = "https://github.com/trifectatechfoundation/libbzip2-rs"
+publish = true
+description = "A memory-safe bzip2 implementation written in rust"
+rust-version = "1.82" # MSRV
+
+[lib]
+name = "bz2_rs" # turns into e.g. `libz_rs.so`
```
Comment (bjorn3) [inline libbzip2-rs-sys-cdylib/Cargo.toml:14]: s/libz_rs.so/libbz2_rs.so/

File: libbzip2-rs-sys-cdylib/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,36 @@
+[package]
+name = "libz-rs-sys-cdylib"
+version = "0.0.0"
+edition = "2021"
+readme = "README.md"
+license = "Zlib"
```
Comment (bjorn3) [inline libbzip2-rs-sys-cdylib/Cargo.toml:6]: BZip2 is not under the Zlib license. The official SPDX license identifier seems to be:

```suggestion
license = "bzip2-1.0.6"
```

File: libbzip2-rs-sys-cdylib/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,36 @@
+[package]
+name = "libz-rs-sys-cdylib"
```
Comment (bjorn3) [inline libbzip2-rs-sys-cdylib/Cargo.toml:2]: Wrong package name.

File: libbzip2-rs-sys-cdylib/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+[lib]
+name = "bz2_rs" # turns into e.g. `libz_rs.so`
+crate-type=["cdylib"]
+
+[features]
+default = ["c-allocator", "libbzip2-rs-sys/stdio",  "libbzip2-rs-sys/std"] # when used as a cdylib crate, use the c allocator
+c-allocator = ["libbzip2-rs-sys/c-allocator"] # by default, use malloc/free for memory allocation
+rust-allocator = ["libbzip2-rs-sys/rust-allocator", "libbzip2-rs-sys/std"] # by default, use the rust global alloctor for memory allocation
+custom-prefix = ["libbzip2-rs-sys/custom-prefix"] # use the LIBZ_RS_SYS_PREFIX to prefix all exported symbols
```
Comment (bjorn3) [inline libbzip2-rs-sys-cdylib/Cargo.toml:21]: s/LIBZ/LIBBZIP2/

File: libbzip2-rs-sys-cdylib/README.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+   Compiling libbzip2-rs-sys v0.0.0 (libbzip2-rs/libbzip2-rs-sys)
+   Compiling libz-rs-sys-cdylib v0.0.0 (libbzip2-rs/libbzip2-rs-sys-cdylib)
+    Finished `release` profile [optimized] target(s) in 0.97s
+> objdump -tT target/release/libbz2_rs.so | grep "BZ2_bzCompressInit"
+000000000002f300 l     F .text	0000000000000441              .hidden _ZN15libbzip2_rs_sys5bzlib22BZ2_bzCompressInitHelp17hac60bda3d983fe05E
+000000000002f2e0 g     F .text	000000000000001a              MY_CUSTOM_PREFIX_BZ2_bzCompressInit
+000000000002f2e0 g    DF .text	000000000000001a  Base        MY_CUSTOM_PREFIX_BZ2_bzCompressInit
+```
+
+### `![no_std]`
```
Comment (bjorn3) [inline libbzip2-rs-sys-cdylib/README.md:71]:
```suggestion
### `#![no_std]`
```

File: libbzip2-rs-sys-cdylib/bzpipe.c [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    else if (argc == 2 && strcmp(argv[1], "-d") == 0) {
+        ret = inf(stdin, stdout);
+        if (ret != BZ_OK)
+            zerr(ret);
+        return ret;
+    }
+
+    /* otherwise, report usage */
+    else {
+        fputs("zpipe usage: zpipe [-d] < source > dest\n", stderr);
```
Comment (bjorn3) [inline libbzip2-rs-sys-cdylib/bzpipe.c:197]:
```suggestion
        fputs("bzpipe usage: bzpipe [-d] < source > dest\n", stderr);
```

File: libbzip2-rs-sys-cdylib/Cargo.toml (hunk 3 above)
Comment (folkertdev) [inline libbzip2-rs-sys-cdylib/Cargo.toml:2]:
```suggestion
name = "libbz2-rs-sys-cdylib"
```

File: libbzip2-rs-sys-cdylib/Cargo.toml (hunk 1 above)
Comment (folkertdev) [inline libbzip2-rs-sys-cdylib/Cargo.toml:14]:
```suggestion
name = "bz2_rs" # turns into e.g. `libbz2_rs.so`
```

File: libbzip2-rs-sys-cdylib/Cargo.toml (hunk 4 above)
Comment (folkertdev) [inline libbzip2-rs-sys-cdylib/Cargo.toml:21]:
```suggestion
custom-prefix = ["libbzip2-rs-sys/custom-prefix"] # use the LIBBZIP2_RS_SYS_PREFIX to prefix all exported symbols
```

File: .github/workflows/checks.yaml [hunk 7]
```diff
@@ -392,3 +392,96 @@ jobs:
...
+          cmp -s Cargo.toml out.txt
+      #      - name: "cdylib: no_std"
+      #        env:
+      #          LD_LIBRARY_PATH: "target/${{matrix.target}}/release/deps"
+      #        working-directory: libbzip2-rs-sys-cdylib
+      #        run: |
+      #          cargo build --release --target ${{matrix.target}} --no-default-features
+      #          cc -o bzpipe_no_std bzpipe_no_std.c target/${{matrix.target}}/release/deps/libbz2_rs.so -I ../
+      #          ./bzpipe_no_std < Cargo.toml | ./bzpipe_no_std -d > out.txt
+      #          cmp -s Cargo.toml out.txt
```
Comment (folkertdev) [inline .github/workflows/checks.yaml:441]: I'm still not sure how this works for zlib-rs and doesn't here

File: libbzip2-rs-sys-cdylib/bzpipe.c [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+        /* done when BZ2_bzDecompress() says it's done */
+    } while (ret != BZ_STREAM_END);
+
+    /* clean up and return */
+    (void)BZ2_bzDecompressEnd(&strm);
+    return ret == BZ_STREAM_END ? BZ_OK : BZ_DATA_ERROR;
+}
+
+/* report a zlib or i/o error */
```
Comment (bjorn3) [inline libbzip2-rs-sys-cdylib/bzpipe.c:150]:
```suggestion
/* report a bzip2 or i/o error */
```

File: libbzip2-rs-sys-cdylib/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+c-allocator = ["libbzip2-rs-sys/c-allocator"] # by default, use malloc/free for memory allocation
+rust-allocator = ["libbzip2-rs-sys/rust-allocator", "libbzip2-rs-sys/std"] # by default, use the rust global alloctor for memory allocation
+custom-prefix = ["libbzip2-rs-sys/custom-prefix"] # use the LIBBZIP2_RS_SYS_PREFIX to prefix all exported symbols
+capi = []
+
+[dependencies]
+libbzip2-rs-sys = { version = "0.0.0", path = "../libbzip2-rs-sys", default-features = false }
+
+[package.metadata.capi.library]
+version = "1.0.9" # the zlib api version we match
```
Comment (bjorn3) [inline libbzip2-rs-sys-cdylib/Cargo.toml:28]:
```suggestion
version = "1.0.9" # the bzip2 api version we match
```

File: libbzip2-rs-sys-cdylib/bzpipe.c [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+#else
+#  define SET_BINARY_MODE(file)
+#endif
+
+#define CHUNK 256
+
+/* Compress from file source to file dest until EOF on source.
+   def() returns BZ_OK on success, BZ_MEM_ERROR if memory could not be
+   allocated for processing, Z_STREAM_ERROR if an invalid compression
+   level is supplied, Z_VERSION_ERROR if the version of zlib.h and the
```
Comment (bjorn3) [inline libbzip2-rs-sys-cdylib/bzpipe.c:23]: This refers to zlib.h and zlib symbols (Z_STREAM_ERROR and Z_VERSION_ERROR).

**Review by bjorn3 — APPROVED** [review]
With the above zlib references fixed.

File: .github/workflows/checks.yaml (hunk 7 above)
Comment (folkertdev) [reply .github/workflows/checks.yaml]: got it, the `[lib]` section of the main crate needs to be removed; that's allright, it is only meant to be used as a crate anyway, building libraries should happen with the cdylib crate.

**Pull request merged**